### ISSUE TITLE
[Bug] Alerting: Malformed/garbled strings with Go functions in Slack webhook

### ISSUE THREAD

**Issue by jacksongoode** [grafana/grafana#63584]
**What happened**:
Our alert templates include some piping and Go formatting like `title` and `reReplaceAll` is formatting some strings but nothing terribly complicated. Sometimes we see totally garbled text that shouldn't appear under any circumstances. This might happen when multiple alerts are fired simultaneously?

For example, `sierra_leone` should have become `Siera Leone` but instead it either appears blank, or becomes something like `sSrrrlLOn`. Same happened with `grafana_state_error` becoming `Graf` or `GrAanaa sAAerAoo`.

**What you expected to happen**:
Expected formatted strings consistently.

**How to reproduce it (as minimally and precisely as possible)**:
I think the culprit might be a ```{{ reReplaceAll "foo" "bar" .Name | title }}``` where this replace is piped into a title? Those are the cases where I have see garbed text.

**Anything else we need to know?**:

**Environment**:
- Grafana version: 9.3.6
- Data source type & version: Varied
- OS Grafana is installed on: Linux
- User OS & Browser: N/A
- Grafana plugins: N/A
- Others: N/A

```[tasklist]

```

**Comment by tonypowa** [comment]
:wave: Thanks for creating this issue, @jacksongoode  

i'm forwarding this issue to the alerting squad
in the meantime, is there a chance the webhook shows in the [logs](https://grafana.com/docs/grafana/v9.0/troubleshooting/) ?

**Comment by grobinson-grafana** [comment]
Hi! Thanks for opening this issue! In Grafana we use [`text/template`](https://pkg.go.dev/text/template), which is the templating language in Go. I am wondering if this could be a bug in `text/template`. Have you been able to reproduce this issue outside of Grafana, for example, just using `text/template`?

**Comment by jacksongoode** [comment]
I would be really surprised if this was a result of the templating language. Especially as it does seem to work when testing the hook. It only fails when many hooks are sent simultaneously.

**Comment by jacksongoode** [comment]
Thanks @tonypowa 
> i'm forwarding this issue to the alerting squad
in the meantime, is there a chance the webhook shows in the [logs](https://grafana.com/docs/grafana/v9.0/troubleshooting/) ?
I'll check on these

**Comment by jacksongoode** [comment]
Hmmm even in the logs, I still get garbled text.

<img width="317" alt="Screen Shot 2023-02-23 at 7 48 42 PM" src="https://user-images.githubusercontent.com/54308792/221087337-53c8417c-a024-4fb8-8d6c-12fedc3c171e.png">

**Comment by jacksongoode** [comment]
I'll just add the templates that are being used for more context:

```
{{ define "myalert" }}
{{ if gt (len .Annotations) 0 }}{{ range .Annotations.SortedPairs }}
*{{ reReplaceAll "[^a-z]" " " .Name | title }}:*{{ if eq .Name "summary" }}
```{{ .Value }}```
{{ else }} {{ .Value }}{{ end }}{{ end }}
{{ else }}{{ if gt (len .Labels) 0 }}
{{ range .Labels.SortedPairs }}
*{{ reReplaceAll "[^a-z]" " " .Name | title }}:* {{ .Value }}{{ end }}{{ end }}
{{ end }}
{{ if gt (len .SilenceURL ) 0 }}<{{ print .SilenceURL "&orgId=1" }}|🔕 Silence>{{ end }}{{ if gt (len .DashboardURL ) 0 }}  |  <{{ .DashboardURL }}|🔗 Dashboard>{{ end }}
{{ end }}
```

```
{{ define "mymessage" }}
  {{ if gt (len .Alerts.Firing) 0 }}
    {{ range .Alerts.Firing }} {{ template "myalert" .}} {{ end }}
  {{ end }}
  {{ if gt (len .Alerts.Resolved) 0 }}
    {{ range .Alerts.Resolved }} {{ template "myalert" .}} {{ end }}
  {{ end }}
{{ end }}
```

```
{{ define "mytitle" }}
  {{ if gt (len .Alerts) 0 }}
    {{ if gt (len .ExternalURL) 0 }}{{ reReplaceAll ".*graphs[^a-z](.*)[^a-z]powerwatch.*" "$1" $.ExternalURL | reReplaceAll "-" " " | title }} - {{ end }}{{ range .Alerts }}{{ .Labels.alertname }}{{ end }}
  {{ end }}
{{ end }}
```

<img width="601" alt="Screen Shot 2023-02-24 at 10 43 18 AM" src="https://user-images.githubusercontent.com/54308792/221264297-7ba9f7b3-e604-4e8f-9f65-edb2efcea51f.png">

**Comment by jacksongoode** [comment]
Just want to follow up with this @grobinson-grafana, is there any other info I can obtain to help debug this?

**Comment by jacksongoode** [comment]
@grobinson-grafana Just to add, 9.4.2 doesn't fix this. A flood of 3/4 alerts at the same time causes messages that are formatted in the template to become garbled in unpredictable ways.

**Comment by gotjosh** [comment]
@jacksongoode the only missing piece we need is what kind of data your alert is outputting.

I also don't understand what:

>It only fails when many hooks are sent simultaneously.

Means? Are you referring to multiple alerts in a single group or...?

**Comment by jacksongoode** [comment]
@grobinson-grafana @armandgrillet Was this resolved in the bugfixes in 9.4.3 or is this slated for 9.4.4?

**Comment by grobinson-grafana** [comment]
Hi! It might be 9.4.5 depending on when the next Alertmanager release is made. Grafana might be able to use the latest main if its stable. @gotjosh what do you think?

**Comment by jacksongoode** [comment]
@grobinson-grafana Was this resolved or is there still an issue?

**Comment by grobinson-grafana** [comment]
Hi! I'm afraid this didn't get released in 9.4.5 as planned, but it will be fixed in the 10.0 beta and 10. I need to talk to the team about how best to backport this to 9.4.x and 9.5.x.

**Comment by grobinson-grafana** [comment]
It will not be until the summer, so will talk with the team how best to backport this to 9.4 and 9.5.

**Comment by grobinson-grafana** [comment]
The pull requests above ^^ will backport this fix to the next minor release of Grafana 9.5.x.

**Comment by jacksongoode** [comment]
Thanks so much! I'll close when they merge.
